### PR TITLE
Windows fix

### LINKED ISSUE #9: fix windows platform missing socket.set_reuse_port(true)
socket.set_reuse_port(true)?;

### LINKED ISSUE #10: [Win/Mac] can't reuse port on Windows/Mac platforms
Examples mdns_server.rs and mdns_query.rs can't run on the same machine for WIndows/Mac platform

### LINKED ISSUE #11: [macos] mdns server does not reply
Hi! I try to open url from chrome, but rust mdns code doesn't reply
there are logs when chrome is asking
```
INFO  [webrtc_mdns::conn] Received new connection from 192.168.43.147:5353
TRACE [webrtc_mdns::conn] recv bytes [0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 10, 109, 121, 45, 115, 101, 114, 118, 105, 99, 101, 5, 108, 111, 99, 97, 108, 0, 0, 1, 128, 1] from 192.168.43.147:5353
TRACE [webrtc_mdns::conn] Found local name: my-service.local. to send answer
TRACE [webrtc_mdns::conn] sent answer from 192.168.43.147 to 224.0.0.251:5353, data: [0, 0, 132, 0, 0, 0, 0, 1, 0, 0, 0, 0, 10, 109, 121, 45, 115, 101, 114, 118, 105, 99, 101, 5, 108, 111, 99, 97, 108, 0, 0, 1, 0, 1, 0, 0, 0, 120, 0, 4, 192, 168, 43, 147]
TRACE [webrtc_mdns::conn] Sent answer to local name: my-service.local. to dst addr 224.0.0.251:5353
TRACE [webrtc_mdns::conn] Parsing has completed
INFO  [webrtc_mdns::conn] Received new connection from 192.168.43.147:5353
TRACE [webrtc_mdns::conn] recv bytes [0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 10, 109, 121, 45, 115, 101, 114, 118, 105, 99, 101, 5, 108, 111, 99, 97, 108, 0, 0, 1, 128, 1] from 192.168.43.147:5353
TRACE [webrtc_mdns::conn] Found local name: my-service.local. to send answer
TRACE [webrtc_mdns::conn] sent answer from 192.168.43.147 to 224.0.0.251:5353, data: [0, 0, 132, 0, 0, 0, 0, 1, 0, 0, 0, 0, 10, 109, 121, 45, 115, 101, 114, 118, 105, 99, 101, 5, 108, 111, 99, 97, 108, 0, 0, 1, 0, 1, 0, 0, 0, 120, 0, 4, 192, 168, 43, 147]
TRACE [webrtc_mdns::conn] Sent answer to local name: my-service.local. to dst addr 224.0.0.251:5353
TRACE [webrtc_mdns::conn] Parsing has completed
INFO  [webrtc_mdns::conn] Received new connection from 127.0.0.1:5353
TRACE [webrtc_mdns::conn] recv bytes [0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 10, 109, 121, 45, 115, 101, 114, 118, 105, 99, 101, 5, 108, 111, 99, 97, 108, 0, 0, 1, 0, 1] from 127.0.0.1:5353
TRACE [webrtc_mdns::conn] Found local name: my-service.local. to send answer
TRACE [webrtc_mdns::conn] sent answer from 127.0.0.1 to 224.0.0.251:5353, data: [0, 0, 132, 0, 0, 0, 0, 1, 0, 0, 0, 0, 10, 109, 121, 45, 115, 101, 114, 118, 105, 99, 101, 5, 108, 111, 99, 97, 108, 0, 0, 1, 0, 1, 0, 0, 0, 120, 0, 4, 127, 0, 0, 1]
TRACE [webrtc_mdns::conn] Sent answer to local name: my-service.local. to dst addr 224.0.0.251:5353
TRACE [webrtc_mdns::conn] Parsing has completed
INFO  [webrtc_mdns::conn] Received new connection from 192.168.43.147:5353
TRACE [webrtc_mdns::conn] recv bytes [0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 10, 109, 121, 45, 115, 101, 114, 118, 105, 99, 101, 5, 108, 111, 99, 97, 108, 0, 0, 1, 0, 1] from 192.168.43.147:5353
TRACE [webrtc_mdns::conn] Found local name: my-service.local. to send answer
TRACE [webrtc_mdns::conn] sent answer from 192.168.43.147 to 224.0.0.251:5353, data: [0, 0, 132, 0, 0, 0, 0, 1, 0, 0, 0, 0, 10, 109, 121, 45, 115, 101, 114, 118, 105, 99, 101, 5, 108, 111, 99, 97, 108, 0, 0, 1, 0, 1, 0, 0, 0, 120, 0, 4, 192, 168, 43, 147]
TRACE [webrtc_mdns::conn] Sent answer to local name: my-service.local. to dst addr 224.0.0.251:5353
TRACE [webrtc_mdns::conn] Parsing has completed
INFO  [webrtc_mdns::conn] Received new connection from 127.0.0.1:5353
TRACE [webrtc_mdns::conn] recv bytes [0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 10, 109, 121, 45, 115, 101, 114, 118, 105, 99, 101, 5, 108, 111, 99, 97, 108, 0, 0, 1, 0, 1] from 127.0.0.1:5353
TRACE [webrtc_mdns::conn] Found local name: my-service.local. to send answer
TRACE [webrtc_mdns::conn] sent answer from 127.0.0.1 to 224.0.0.251:5353, data: [0, 0, 132, 0, 0, 0, 0, 1, 0, 0, 0, 0, 10, 109, 121, 45, 115, 101, 114, 118, 105, 99, 101, 5, 108, 111, 99, 97, 108, 0, 0, 1, 0, 1, 0, 0, 0, 120, 0, 4, 127, 0, 0, 1]
TRACE [webrtc_mdns::conn] Sent answer to local name: my-service.local. to dst addr 224.0.0.251:5353
TRACE [webrtc_mdns::conn] Parsing has completed
INFO  [webrtc_mdns::conn] Received new connection from 192.168.43.147:5353
TRACE [webrtc_mdns::conn] recv bytes [0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 10, 109, 121, 45, 115, 101, 114, 118, 105, 99, 101, 5, 108, 111, 99, 97, 108, 0, 0, 1, 0, 1] from 192.168.43.147:5353
TRACE [webrtc_mdns::conn] Found local name: my-service.local. to send answer
TRACE [webrtc_mdns::conn] sent answer from 192.168.43.147 to 224.0.0.251:5353, data: [0, 0, 132, 0, 0, 0, 0, 1, 0, 0, 0, 0, 10, 109, 121, 45, 115, 101, 114, 118, 105, 99, 101, 5, 108, 111, 99, 97, 108, 0, 0, 1, 0, 1, 0, 0, 0, 120, 0, 4, 192, 168, 43, 147]
TRACE [webrtc_mdns::conn] Sent answer to local name: my-service.local. to dst addr 224.0.0.251:5353
TRACE [webrtc_mdns::conn] Parsing has completed
```

and this is log when i am running cmd ```dns-sd -q my-service.local```
```
INFO  [webrtc_mdns::conn] Received new connection from 192.168.43.147:5353
TRACE [webrtc_mdns::conn] recv bytes [0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 10, 109, 121, 45, 115, 101, 114, 118, 105, 99, 101, 5, 108, 111, 99, 97, 108, 0, 0, 1, 128, 1] from 192.168.43.147:5353
TRACE [webrtc_mdns::conn] Found local name: my-service.local. to send answer
TRACE [webrtc_mdns::conn] sent answer from 192.168.43.147 to 224.0.0.251:5353, data: [0, 0, 132, 0, 0, 0, 0, 1, 0, 0, 0, 0, 10, 109, 121, 45, 115, 101, 114, 118, 105, 99, 101, 5, 108, 111, 99, 97, 108, 0, 0, 1, 0, 1, 0, 0, 0, 120, 0, 4, 192, 168, 43, 147]
TRACE [webrtc_mdns::conn] Sent answer to local name: my-service.local. to dst addr 224.0.0.251:5353
TRACE [webrtc_mdns::conn] Parsing has completed
INFO  [webrtc_mdns::conn] Received new connection from 192.168.43.147:5353
TRACE [webrtc_mdns::conn] recv bytes [0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 10, 109, 121, 45, 115, 101, 114, 118, 105, 99, 101, 5, 108, 111, 99, 97, 108, 0, 0, 1, 128, 1] from 192.168.43.147:5353
TRACE [webrtc_mdns::conn] Found local name: my-service.local. to send answer
TRACE [webrtc_mdns::conn] sent answer from 192.168.43.147 to 224.0.0.251:5353, data: [0, 0, 132, 0, 0, 0, 0, 1, 0, 0, 0, 0, 10, 109, 121, 45, 115, 101, 114, 118, 105, 99, 101, 5, 108, 111, 99, 97, 108, 0, 0, 1, 0, 1, 0, 0, 0, 120, 0, 4, 192, 168, 43, 147]
TRACE [webrtc_mdns::conn] Sent answer to local name: my-service.local. to dst addr 224.0.0.251:5353
TRACE [webrtc_mdns::conn] Parsing has completed
INFO  [webrtc_mdns::conn] Received new connection from 127.0.0.1:5353
TRACE [webrtc_mdns::conn] recv bytes [0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 10, 109, 121, 45, 115, 101, 114, 118, 105, 99, 101, 5, 108, 111, 99, 97, 108, 0, 0, 1, 0, 1] from 127.0.0.1:5353
TRACE [webrtc_mdns::conn] Found local name: my-service.local. to send answer
TRACE [webrtc_mdns::conn] sent answer from 127.0.0.1 to 224.0.0.251:5353, data: [0, 0, 132, 0, 0, 0, 0, 1, 0, 0, 0, 0, 10, 109, 121, 45, 115, 101, 114, 118, 105, 99, 101, 5, 108, 111, 99, 97, 108, 0, 0, 1, 0, 1, 0, 0, 0, 120, 0, 4, 127, 0, 0, 1]
TRACE [webrtc_mdns::conn] Sent answer to local name: my-service.local. to dst addr 224.0.0.251:5353
TRACE [webrtc_mdns::conn] Parsing has completed
INFO  [webrtc_mdns::conn] Received new connection from 192.168.43.147:5353
TRACE [webrtc_mdns::conn] recv bytes [0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 10, 109, 121, 45, 115, 101, 114, 118, 105, 99, 101, 5, 108, 111, 99, 97, 108, 0, 0, 1, 0, 1] from 192.168.43.147:5353
TRACE [webrtc_mdns::conn] Found local name: my-service.local. to send answer
TRACE [webrtc_mdns::conn] sent answer from 192.168.43.147 to 224.0.0.251:5353, data: [0, 0, 132, 0, 0, 0, 0, 1, 0, 0, 0, 0, 10, 109, 121, 45, 115, 101, 114, 118, 105, 99, 101, 5, 108, 111, 99, 97, 108, 0, 0, 1, 0, 1, 0, 0, 0, 120, 0, 4, 192, 168, 43, 147]
TRACE [webrtc_mdns::conn] Sent answer to local name: my-service.local. to dst addr 224.0.0.251:5353
TRACE [webrtc_mdns::conn] Parsing has completed
INFO  [webrtc_mdns::conn] Received new connection from 127.0.0.1:5353
TRACE [webrtc_mdns::conn] recv bytes [0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 10, 109, 121, 45, 115, 101, 114, 118, 105, 99, 101, 5, 108, 111, 99, 97, 108, 0, 0, 1, 0, 1] from 127.0.0.1:5353
TRACE [webrtc_mdns::conn] Found local name: my-service.local. to send answer
TRACE [webrtc_mdns::conn] sent answer from 127.0.0.1 to 224.0.0.251:5353, data: [0, 0, 132, 0, 0, 0, 0, 1, 0, 0, 0, 0, 10, 109, 121, 45, 115, 101, 114, 118, 105, 99, 101, 5, 108, 111, 99, 97, 108, 0, 0, 1, 0, 1, 0, 0, 0, 120, 0, 4, 127, 0, 0, 1]
TRACE [webrtc_mdns::conn] Sent answer to local name: my-service.local. to dst addr 224.0.0.251:5353
TRACE [webrtc_mdns::conn] Parsing has completed
INFO  [webrtc_mdns::conn] Received new connection from 192.168.43.147:5353
TRACE [webrtc_mdns::conn] recv bytes [0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 10, 109, 121, 45, 115, 101, 114, 118, 105, 99, 101, 5, 108, 111, 99, 97, 108, 0, 0, 1, 0, 1] from 192.168.43.147:5353
TRACE [webrtc_mdns::conn] Found local name: my-service.local. to send answer
TRACE [webrtc_mdns::conn] sent answer from 192.168.43.147 to 224.0.0.251:5353, data: [0, 0, 132, 0, 0, 0, 0, 1, 0, 0, 0, 0, 10, 109, 121, 45, 115, 101, 114, 118, 105, 99, 101, 5, 108, 111, 99, 97, 108, 0, 0, 1, 0, 1, 0, 0, 0, 120, 0, 4, 192, 168, 43, 147]
TRACE [webrtc_mdns::conn] Sent answer to local name: my-service.local. to dst addr 224.0.0.251:5353
TRACE [webrtc_mdns::conn] Parsing has completed
```
i make a dump raw_answer var to the logs in _conn/mod.rs:411_
```rust
    socket.send_to(&raw_answer, dst_addr).await?;
    log::trace!("sent answer from {} to {}, data: {:?}", dst, dst_addr, raw_answer);
```
I can't see mdns response messages in Wireshark.. only mdns queries from chrome and cmd tool
 
I will try to investigate an issue!)

### PR DESCRIPTION
**Pull request by metaclips**
closes #10
closes #11

### CODE REVIEW

File: examples/mdns_server.rs [hunk 1]
```diff
@@ -47,13 +45,14 @@ async fn main() -> Result<(), Error> {
         );
 
     let matches = app.clone().get_matches();
+    let server = matches.value_of("server").unwrap();
```
Comment (rainliu) [inline examples/mdns_server.rs:48]: any difference moving line 55 to here?

File: examples/mdns_server.rs [hunk 2]
```diff
@@ -65,18 +64,8 @@ async fn main() -> Result<(), Error> {
     )
     .unwrap();
 
-    let running = Arc::new(AtomicBool::new(true));
-    let r = running.clone();
-
-    ctrlc::set_handler(move || {
```
Comment (rainliu) [inline examples/mdns_server.rs:71]: let's remove ctrlc = "3.1.8" crate from Cargo.toml

File: examples/mdns_server.rs [hunk 3]
```diff
@@ -34,7 +32,7 @@ async fn main() -> Result<(), Error> {
             Arg::with_name("server")
                 .required_unless("FULLHELP")
                 .takes_value(true)
-                .default_value(DEFAULT_DEST_ADDR)
+                .default_value("0.0.0.0:5353")
```
Comment (rainliu) [inline examples/mdns_server.rs:35]: so, we can't use "224.0.0.251:5353" for multicast listening? Is that limitation in Windows or Mac?

File: examples/mdns_server.rs (hunk 2 above)
Comment (metaclips) [reply examples/mdns_server.rs:71]: Fixed.

File: examples/mdns_server.rs (hunk 3 above)
Comment (metaclips) [reply examples/mdns_server.rs:35]: We use the INADDR_ANY `0.0.0.0` so as to bind to all interfaces both on windows and unix. Specifying `224.0.0.251:5353` wont bind as it is not a valid address to any interface.

File: examples/mdns_server.rs (hunk 1 above)
Comment (metaclips) [reply examples/mdns_server.rs:48]: None at all. Must have made the change while testing.

**Comment by rainliu** [comment]
@metaclips , I tested your PR in MacOS.

cargo run --color=always --package webrtc-mdns --example mdns_server
cargo run --color=always --package webrtc-mdns --example mdns_query

However, there are two issues:
1). mdns_server can't be stop by "ctrl-c".
2). mdns_query won't exit the program after received answer, looks like something wrong in "server.close().await.unwrap();"

**Comment by metaclips** [comment]
> @metaclips , I tested your PR in MacOS.
> 
> cargo run --color=always --package webrtc-mdns --example mdns_server
> cargo run --color=always --package webrtc-mdns --example mdns_query
> 
> However, there are two issues:
> 1). mdns_server can't be stop by "ctrl-c" immediately after mdns_query sent the query.
> 2). mdns_query won't exit the program after received answer, looks like something wrong in "server.close().await.unwrap();"

This is an issue with recv_from socket function. Since a timeout was not specified, it blocks and only exits on the next recv.
I added a timeout of 2 seconds to mitigate this.

**Comment by rainliu** [comment]
don't use timeout, instead, use cancel channel. So when close mdns_server, it sends cancel signal to cancel channel to break the loop.

**Comment by metaclips** [comment]
`set_read_timeout` docs states that
```rust
/// Set value for the `SO_RCVTIMEO` option on this socket.
///
/// If `timeout` is `None`, then `read` and `recv` calls will block
/// indefinitely.
```
If no timeout is added, `recv_from` blocks here https://github.com/metaclips/mdns/blob/90cfc8341aa0cd51e55f4966ffc650ec589e3952/src/conn/mod.rs#L257 thereby blocking the close signal to be selected. I have indicated the channel to be non-blocking.
More on the issue was discussed here https://github.com/tokio-rs/tokio/issues/3084

**Comment by rainliu** [comment]
Got it.

**Comment by rainliu** [comment]
just test it on Mac OS, works fine.

**Pull request merged**